### ISSUE TITLE
Add BoxConstraints.screen()

### ISSUE THREAD

**Issue by feinstein** [flutter/flutter#57782]
I tried to place a `Scaffold` with a `NestedScrollView` inside a `AnimatedCrossFade` that's inside a `FutureBuilder` and I got several rendering errors saying:

(`FutureBuilder` > `AnimatedCrossFade` > `Scaffold` > `NestedScrollView`)

```
RenderCustomMultiChildLayoutBox object was given an infinite size during layout.

This probably means that it is a render object that tries to be as big as possible, but it was put inside another render object that allows its children to pick their own size.
```

I could fix it using:

```dart
  AnimatedCrossFade(
...
    secondChild: ConstrainedBox(
      constraints: BoxConstraints(maxHeight: MediaQuery.of(context).size.height),
      child: Scaffold(
        body: NestedScrollView(
...
```

Since wanting to constrain by the screen dimensions should be a common thing I think it would be nice to have a `BoxConstraints.screen()` that would be just a `MediaQuery.of(context).size` call internally.

**Comment by britannio** [comment]
Does `SizedBox.expand()` work?

**Comment by feinstein** [comment]
Nope.

**Comment by matthew-carroll** [comment]
I don't think this issue is specific to the complexity described in the original post. I think I'm running into the situation with a much simpler scenario:

```
AnimatedCrossFade(
      crossFadeState:CrossFadeState.showFirst,
      duration: const Duration(milliseconds: 500),
      firstChild: Stack(
        children: [
          Container(
            width: double.infinity,
            height: double.infinity,
            color: Colors.red,
          ),
        ],
      ),
      secondChild: Stack(
        children: [
          Container(
            width: double.infinity,
            height: double.infinity,
            color: Colors.blue,
          ),
        ],
      ),
    );
```

The above code essentially wraps a fullscreen `Stack` and `Container` with an `AnimatedCrossFade`. Without the `AnimatedCrossFade`, either of those `Stack`s render just fine. But with the `AnimatedCrossFade` I get the error reported with the original post.

I believe the root cause of the error is here:
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/animated_cross_fade.dart#L221

At that link you'll see the following:
```
static Widget defaultLayoutBuilder(Widget topChild, Key topChildKey, Widget bottomChild, Key bottomChildKey) {
    return Stack(
      clipBehavior: Clip.none,
      children: <Widget>[
        Positioned(
          key: bottomChildKey,
          left: 0.0,
          top: 0.0,
          right: 0.0,
          child: bottomChild,
        ),
        Positioned(
          key: topChildKey,
          child: topChild,
        ),
      ],
    );
  }
```

Notice that the layout tree for the `AnimatedCrossFade` explicitly leaves out the `bottom` position. This effectively overrides the natural sizing of the `AnimatedCrossFade` and instead requires that the top child impose an intrinsic height.

I'm not sure what the basis was for this decision, but I don't think that it's generally appropriate. I think one should be able to take a widget in whatever layout it belongs to and then wrap that widget in an `AnimatedCrossFade` without needing to know the specifics of the bounding constraints. Or, at least, that approach should be an option when using `AnimatedCrossFade`.

PS - the notion of saying "I want to be as big as the screen" is not something I'd recommend as a solution to this. First, I don't think we want developers to be in the habit of referencing the global screen. This breaks composition and can only possibly makes things worse longer term as Flutter goes to platforms with very different screen dynamics. Second, "screen" is ambiguous. Do you mean the physical screen? Do you mean the safe area of the screen? Third, there could easily be cases where one would like to abide by the incoming constraints, even when those constraints do not represent the whole screen.

**Comment by matthew-carroll** [comment]
I just realized that `AnimatedSwitcher` is also available:
https://api.flutter.dev/flutter/widgets/AnimatedSwitcher-class.html

Perhaps the intention was that `AnimatedSwitcher` uses incoming constraints whereas `AnimatedCrossFade` does not. However, if that was the intention, I still think there are inconsistencies. For example, `AnimatedSwitcher` can switch between any number of children but `AnimatedCrossFade` cannot. So if one is intended to be an intrinsic-sized version of the other then I think there are some modifications that need to take place to one, the other, or both.

**Comment by feinstein** [comment]
I have had lots of issues with `AnimatedCrossFade`, if you search my previous issues you find many of them, my general feeling is that `AnimatedCrossFade` is badly implemented.

That being said, my point with `screen` is to have a way to declare bounds where those bounds have a limit so we can avoid conflicts of `infinity`. Also, in my understanding, screen  is the device screen so if someone what's the visible part he would still have to include the `SafeArea`. But my proposition is more on the general idea than on the specific mechanism that I suggested.

**Comment by sooxt98** [comment]
I cant crossfade between two scaffold ...
